### ISSUE TITLE
Please improve tagging capability and SDK docs

### ISSUE THREAD

**Issue by kristianmandrup** [Dynatrace/OneAgent-SDK-for-dotnet#5]
Currently resorting to looking at the tests or code samples

But is all just "dummy" stuff. 

Had hoped I had found sth useful. Where can I find docs of what is currently available (as per 1.2) and how to use it. More interested in the `IncomingRemoteCallTracer`

Had hope we could somehow tag with `applicationId` or sth useful

`incomingRemoteCallTracer.SetDynatraceStringTag(applicationId);`

Turns out this is limited to connecting outgoing and incoming request tracing via "bonding" on a DynaTrace generated tag from what I can gather from sample app.

Really hoping for sth like custom request attributes ASAP: https://www.dynatrace.com/news/blog/capture-any-request-attributes-using-oneagent-sdk/

My boss wants some progress on this task, but I am completely blocked by current SDK limitations and lack of docs. 

Hope you can give some advice for how to proceed. Perhaps using a DynaTrace HTTPS/REST API directly?  Is this what the SDK is using internally or is it a proprietary protocol?

**Comment by kristianmandrup** [comment]
Hmm, looks like it is somewhat possible via: https://www.dynatrace.com/support/help/monitor/transactions-and-services/request-attributes/how-do-i-capture-request-attributes-based-on-dotnet-method-arguments/

But that would require us to manually change and extend all our controller methods in our API gateway!? Not really feasible at this point.

**Comment by arminru** [comment]
Hi Kristian!

Regarding the DummyImpls:
The SDK is based on our OneAgent, which is installed separately and provides the implementation for collecting data, creating PurePaths and sending those to our server. The DummyImpls which you can find in the repo are in place when no OneAgent is installed and basically provide a no-op implementation so that your application works without a OneAgent dependency as well. The Dummy*Tests test this DummyImpls to assure they do not cause any crashes (e.g. when ITracer.Trace(null)) is called, do not swallow any exceptions and do execute passed Actions/Funcs as expected.
For proper, useful samples please refer to the ones in our readme and to our sample project in the samples/ subfolder.

The available features are listed in the readme, together with the SDK version in which they were introduced:
https://github.com/Dynatrace/OneAgent-SDK-for-dotnet#features
For future features please refer to our roadmap in this thread:
https://answers.dynatrace.com/spaces/483/dynatrace-product-ideas/idea/198106/planned-features-for-oneagent-sdk.html
Please note that the SDK verson 1.2 is not released yet. We'll merge it on master, announce a Github release and publish the package to nuget.org once OneAgent it requires (version 1.161) is released.

Regarding tagging please refer to our general OneAgent SDK documentation: https://github.com/Dynatrace/OneAgent-SDK#tracers
In order to connect a service call between two processes, you will have to use `Outgoing*Tracer.GetDynatraceStringTag` (or `GetDynatraceByteTag`) to generate a tag on the outgoing site, transport it along the call and set it on the incoming site using `Incoming*Tracer.SetDynatraceStringTag` (or `SetDynatraceByteTag`). Please note that you cannot use any custom value, like your `applicationId`, here.
If you want to connect calls between services/processes, using these methods would be the way to go.
If you want to be able to filter your requests/services according to certain meta data in Dynatrace (like your `applicationId`) custom request attributes are the way to go. Feel free to add a comment on our roadmap (https://answers.dynatrace.com/spaces/483/dynatrace-product-ideas/idea/198106/planned-features-for-oneagent-sdk.html) so that Product Management is able to prioritize it according to customer needs.

In order to be able to give you advice it would be helpful if you could illustrate your use case further. Unfortunately I am not really sure what you want to achieve, yet. The communication between the SDK/OneAgent and our server is proprietary.

**Comment by kristianmandrup** [comment]
Thanks @arminru 

We basically need to communicate Application ID and context meta data to Dynatrace Collector.
Custom Request Attributes are exactly what we need. 

PS: I could not find any information on how to set `ApplicationID` from any type of data source.

**Comment by arminru** [comment]
What kind of Application ID do you mean and where would you expect it to show up in Dynatrace or what would you like to do with this ID?

**Comment by kristianmandrup** [comment]
By `ApplicationId` I mean the Web Application Id mentioned here https://www.dynatrace.com/support/help/monitor/transactions-and-services/introduction/how-does-dynatrace-detect-and-name-services/

**Comment by kristianmandrup** [comment]
Also, would Dynatrace openKit, be a valid option for a web API server? 

If so, is it compatibilty with Dynatrace 7.0?  https://www.dynatrace.com/support/help/deploy-dynatrace/other-deployment-options/openkit/instrument-your-application-using-dynatrace-openkit/

Thanks

**Comment by arminru** [comment]
Ah I see, thanks.
Would you like to override an automatically detected Web application ID or is none detected and you would like to provide one manually? Which web service framework are you using?

**Comment by kristianmandrup** [comment]
Yes, would be perfect if that was possible. 

As far as I know our Web application Id is null and void, empty. We have no idea how to make DynaTrace detect it, how it determines it, what needs to be set.

**Comment by arminru** [comment]
Would the approach mentioned at https://www.dynatrace.com/support/help/monitor/transactions-and-services/introduction/how-does-dynatrace-detect-and-name-services/ be suitable for you?

> For other technologies, or to provide a default name, you can use the environment variable `DT_APPLICATIONID=<name>`. This won't override the naming options mentioned above—it serves as a default for instances when no application ID is available.

**Comment by kristianmandrup** [comment]
We are using .NET MVC 5.2.3. We can NOT rely on an environment variable as our Web application Id is "virtual", ie. we have one API gateway serving multiple apps (fan in) and then calling multiple backend services (fan out). So for the gateway, we need to specify the Web application Id (or custom Request Attributes) depending on metadata present in the incoming request (as far as I can see it).

Your thoughts?

**Comment by AlramLechnerDynatrace** [comment]
@kristianmandrup 

> Also, would Dynatrace openKit, be a valid option for a web API server?
as you need instrumenting code on server side, OpenKit seems not the right option for you. ( see:
https://github.com/Dynatrace/openkit-dotnet#what-you-cannot-do-with-the-openkit)

> If so, is it compatibilty with Dynatrace 7.0?
just to be sure: even OpenKit works with Dynatrace AppMon, the OneAgentSDK does not. OneAgent SDK is working with Dynatrace only.

> As far as I know our Web application Id is null and void, empty. We have no idea how to make DynaTrace detect it, how it determines it, what needs to be set.
If agent's internal WebRequestSensor doesn't detect ApplicationId in the right way: please open an issue in our support portal: https://support.dynatrace.com/supportportal/. issues about the agents internal sensor in the OneAgent SDK project are out of scope.

something you should also aware: you can't change the applicationId detected by the agents internal sensor with the OneAgent SDK.

setting your applicationId as custom request attribute (https://github.com/Dynatrace/OneAgent-SDK#scav) seems to be an valid workaorund for you. but - as already mentioned here - custom request attributes are not available in .net now (good news: it is on the roadmap - but for now i can't provide any timeframe when it will be available). 
to push this feature to be the next one, feel free to leave a comment on the roadmap page: https://answers.dynatrace.com/spaces/483/dynatrace-product-ideas/idea/198106/planned-features-for-oneagent-sdk.html and/or reach out Sonja Chevre (https://answers.dynatrace.com/users/3366/view.html). She is responsible PM for the OneAgent SDK.

please let me know, if there are any other open questions. i'll close the issue for now, as there is nothing we can do from the SDK side.
BR Alram

**Comment by kristianmandrup** [comment]
We will be using custom request attributes going forward, for now via `methods` nd later via the SDK when they are available. Thank you so much.

**Comment by AlramLechnerDynatrace** [comment]
Great - thanks for your feedback!
